File: matlab/prediction_correction/proj_corr.m
function proj_corr(A,epss,lamb0,K,tol,tho)
    vp = eig(A);
    taille = size(vp);
    taille = taille(1);
    i=0;
    z1_old = 0;
    [n, l] = size(A);
    theta0 = epss;
    d0 = 1i;
    z = lamb0 + theta0*d0;
    g = min(svd(z*eye(n)-A));
    k = 0;
    u_min = [];
    v_min = [];
    while(abs(g-epss)/epss>tol)
        z1_old = z;
        [u,s,v] = svd(z1_old*eye(n)-A);
        s = diag(s);
        [s,i] = min(s);
        u_min = u(1:n,i:i);
        v_min = v(1:n,i:i);
        z = z - ((s - epss)/real(conj(d0)*ctranspose(v_min)*u_min))*d0;
        g = min(svd(z*eye(n)-A));
    end
    z1 = z
    while(k<K)
        r = 1i*(transpose(v_min)*u_min)/abs(ctranspose(v_min)*u_min);
        z = z+tho*r;
        k = k+1;
        [u,s,v] = svd(z*eye(n)-A);
        s = diag(s);
        [s,i] = min(s);
        u_min = u(1:n,i:i);
        v_min = v(1:n,i:i);
        z = z - ((s - epss)/(ctranspose(u_min)*v_min))
        zx(k) = real(z);
        zy(k) = imag(z);
        real(z1)
        if(abs(z-z1) < 0.001*abs(z1))
            break;
        end
    end
    plot(zx,zy);
end

File: python/proj_corr/proj_corr.py
import numpy as np
import matplotlib.pyplot as plt

def proj_corr(A,epss,K,tol) :
    vp = np.linalg.eigvals(A);
    taille = np.size(vp);
    for i in range(taille):
        #STEP 0: Compute the first point z1
        z1_old = 0;
        n, l = np.shape(A);
        theta0 = epss;
        d0 = 1j;
        lamb0 = vp[i];
        zx = [];
        zy = [];
        z = lamb0 + theta0*d0;
        u,g,v = np.linalg.svd(z*np.eye(n)-A);
        g = g[-1]
        k = 0;
        u_min = [];
        v_min = [];
        while(abs(g-epss)/epss>tol):
            z1_old = z;
            u,s,v = np.linalg.svd(z1_old*np.eye(n)-A);
            s =s[-1];
            u= u.T
            u_min = u[-1]
            v_min = v[-1]
            v_min = np.conj(v_min)
            z = z - ((s - epss)/(np.conj(d0)*(np.vdot(v_min,u_min))).real)*d0
            g = np.linalg.svd(z*np.eye(n)-A,compute_uv = 0);
            g = g[-1]
        z1 = z 
        #END STEP 0
        while(k<K):
            #STEP 1: Prediction
            tho = np.min(0.1,(0.5)*abs(z - vp[i]))
            r = 1j*(np.vdot(np.conj(v_min),u_min))/abs(np.vdot(v_min,u_min))
            z = z+tho*r
            #STEP 2: Correction
            u,s,v = np.linalg.svd(z*np.eye(n)-A)
            s = s[-1]
            u= u.T
            u_min = u[-1]
            v_min = v[-1]
            v_min = np.conj(v_min)
            z = z - ((s - epss)/(np.vdot(u_min,v_min)))
            zx.append(z.real);
            zy.append(z.imag);
            if (k == 2 or k == 7):
                print("zx:",zx)
                print("zy:",zy)
            if(abs(z-z1) < 0.001*abs(z1)):
                break;
            k = k+1;
        plt.plot(zx,zy);
    plt.show()

A = np.array([[1, 1, 1, 1],[2, 60, 2, 2],[1, 1, 20, 1],[1, 2, 5, 20*1j]])
proj_corr(A,0.3,10000,0.0001)
